ocaml/xapi/vm_sysprep.ml

```diff
@@ -0,0 +1,249 @@
+
+(** Find the VBD for the CD drive on [vm] *)
+let find_cdr_vbd ~__context ~vm =
+  let vbds = Db.VM.get_VBDs ~__context ~self:vm in
+  let vbds' =
+    List.map (fun self -> (self, Db.VBD.get_record ~__context ~self)) vbds
+  in
+  let is_cd (_rf, rc) =
+    let open API in
+    rc.vBD_type = `CD && rc.vBD_empty
+  in
+  let uuid = Db.VM.get_uuid ~__context ~self:vm in
+  match List.filter is_cd vbds' with
+  | [] ->
+      fail "can't find CDR for VM %s" uuid
+  | [(rf, rc)] ->
+      debug "%s: for VM %s using VBD %s" __FUNCTION__ uuid rc.API.vBD_uuid ;
+      rf
+  | (rf, rc) :: _ ->
+      debug "%s: for VM %s using VBD %s" __FUNCTION__ uuid rc.API.vBD_uuid ;
+      warn "%s: for VM %s found additions VBDs" __FUNCTION__ uuid ;
+      rf
+
+(** Find the VDI that contains the unattend.xml based on its name. This
+    should be unique *)
+let find_vdi ~__context ~label =
+  match Db.VDI.get_by_name_label ~__context ~label with
+  | [] ->
+      internal_error "%s: can't find VDI for %s" __FUNCTION__ label
+  | [vdi] ->
+      vdi
+  | vdi :: _ ->
+      warn "%s: more than one VDI with label %s" __FUNCTION__ label ;
+      vdi
+
+(** notify the VM with [domid] to run sysprep and where to find the
+    file. *)
+let trigger ~domid =
+  let open Ezxenstore_core.Xenstore in
+  let control = Printf.sprintf "/local/domain/%Ld/control/sysprep" domid in
+  with_xs (fun xs ->
+      xs.Xs.write (control // "filename") "D://unattend.xml" ;
+      Thread.delay 5.0 ;
+      xs.Xs.write (control // "action") "sysprep" ;
+      debug "%s: notified domain %Ld" __FUNCTION__ domid ;
+      let rec wait n =
+        match (n, xs.Xs.read (control // "action")) with
+        | _, "running" ->
+            "running"
+        | n, action when n < 0 ->
+            action
+        | _, _ ->
+            Thread.delay 1.0 ;
+            wait (n - 1)
+      in
+      (* wait up to 5 iterations for runnung to appear or report whatever
+         is the status at the end *)
+      wait 5
+  )
+
+(* This function is executed on the host where [vm] is running *)
+let sysprep ~__context ~vm ~unattend =
+  let open Ezxenstore_core.Xenstore in
+  debug "%s" __FUNCTION__ ;
+  if not !Xapi_globs.vm_sysprep_enabled then
+    fail "Experimental VM.sysprep API call is not enabled" ;
+  let vm_uuid = Db.VM.get_uuid ~__context ~self:vm in
+  let domid = Db.VM.get_domid ~__context ~self:vm in
+  let control = Printf.sprintf "/local/domain/%Ld/control" domid in
+  if domid <= 0L then
+    fail " VM %s is not running" __FUNCTION__ vm_uuid ;
+  if String.length unattend > 32 * 1024 then
+    fail "%s: provided file for %s larger than 32KiB" __FUNCTION__ vm_uuid ;
+  with_xs (fun xs ->
+      match xs.Xs.read (control // "feature-sysprep") with
+      | "1" ->
+          debug "%s: VM %s supports sysprep" __FUNCTION__ vm_uuid
+      | _ ->
+          fail "VM %s does not support sysprep" vm_uuid
+      | exception _ ->
+          fail "VM %s does not support sysprep" vm_uuid
+  ) ;
```

Comment:
This could eventually be replaced by allowed_operations checking (see `check_op_for_feature` in `xapi_vm_lifecycle.ml`).

Reply:
I think this test is good because it checks against the driver directly.